### PR TITLE
Add MultiEnum renderer for enums and oneOf

### PR DESCRIPTION
**Pull request by AlexandraBuzila**
Render an anyOf array of consts as a list of checkboxes. The checkbox is
checked if the array item it represents is present in the data.

### CODE REVIEW

File: packages/material/src/complex/MaterialEnumArrayControl.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        schema => hasType(schema, 'array') && !Array.isArray(schema.items)
+      ),
+      schemaSubPathMatches('items', schema => {
+        return schema.anyOf !== undefined && schema.anyOf.length > 0;
+      })
+    )
+  )
+);
+
+export default MaterialEnumArrayControl;
```
Comment (sdirix) [inline packages/material/src/complex/MaterialEnumArrayControl.tsx:115]: To be consistent with other controls we should default export the wrapped version

File: packages/material/src/index.ts [hunk 2]
```diff
@@ -171,6 +174,10 @@ export const materialRenderers: JsonFormsRendererRegistryEntry[] = [
   {
     tester: materialAnyOfStringOrEnumControlTester,
     renderer: MaterialAnyOfStringOrEnumControl
+  },
+  {
+    tester: materialEnumArrayControlTester,
+    renderer: withJsonFormsControlProps(MaterialEnumArrayControl)
```
Comment (sdirix) [inline packages/material/src/index.ts:180]: Don't wrap here, but use the already wrapped imported version

File: packages/material/src/complex/MaterialEnumArrayControl.tsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  return (
+    <Hidden xlUp={!visible}>
+      <FormControl component='fieldset'>
+        <FormGroup row>
+          {uiData.map((value: any, index: number) => {
+            const option = options[index];
+            const optionPath = Paths.compose(path, `${index}`);
+            return (
+              <FormControlLabel
+                id={option.const}
```
Comment (sdirix) [inline packages/material/src/complex/MaterialEnumArrayControl.tsx:70]: Careful: The value of `const` can be anything! It doesn't need to be a string, but could for example be an `object`, `array`, etc. Please see `mapStateToOneOfEnumProps` where we already had to handle the case.

File: packages/material/src/complex/MaterialEnumArrayControl.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+      // option was added
+      newData.push(option);
+    } else {
+      // option was removed
+      const indexInData = newData.indexOf(option);
+      newData.splice(indexInData, 1);
+    }
+
+    handleChange(path, newData);
+  };
```
Comment (sdirix) [inline packages/material/src/complex/MaterialEnumArrayControl.tsx:56]: If you would use `withJsonFormsArrayControlProps` you would get functions to add and remove items to the array for free, i.e. `addItem`, `removeItems`, `moveUp` and `moveDown` would get injected. `schema` would also already point to the `items` object.

**Review by sdirix — CHANGES_REQUESTED** [review]
In general I like it very much. However there are some things I would like to see differently.

Currently the schema tested for looks like this
```js
"property": {
  "type": "array",
  "items": {
    "anyOf": [
      { },
    ]
  },
},
 ```
 which doesn't really fit to the use case:

- The schema allows the same entry multiple times, but our control only handles them uniquely (the control therefore misbehaves if the same entry is there multiple times)
- `anyOf` stands for that **at least** one item matches. However we're handling it like "enums", which are unique. So it should be `oneOf`.
- To be consistent with our other `oneOfEnum` controls we should check whether the `items` are `oneOf` and contain a `const` in each entry.
- Ideally we should also be able to handle `enum` content

I would prefer to support the following schemas:
```js
"property": {
  "type": "array",
  "items": {
    "oneOf": [
      {
        "const": <any>,
        "title": "mylabel"
      },
    ]
  },
  "uniqueItems": true
},
```
and
```js
"property": {
  "type": "array",
  "items": {
    "type": "string",
    "enum": ["a", "b", "c"]
  },
  "uniqueItems": true
},
```

So we check for

 - type array with unique items
 - the items are either `enum` or `oneOf` containing `const`
 
 ---
 
 Please also add an example to the example app.

**Review by sdirix — CHANGES_REQUESTED** [review]
Additionally it would be great when we don't add all the logic into the renderer. Instead I would prefer a new `mapStateToMultiEnumProps` and `mapDispatchToMultiEnumProps`.

- `mapStateToMultiEnumProps`: extracting the `items` options from `oneOf` or `enum`, similar to `mapStateToEnumPros` and `mapStateToOneOfEnumProps`
- `mapDispatchToMultiEnumProps`: providing a `addItem` and `removeItem` function which can be used for toggling. Using this we can also remove the `useRef` workaround

By extracting the logic from the renderer we can easily write a multi-select enum with a different renderer (e.g. Material UI's multi select dropdown) as well as more easily add the same renderer in the other renderer sets too.

**Comment by AlexandraBuzila** [comment]
I updated the commit to include you comments, Stefan. The renderer now supports both `oneOf` and `enum` items.

**Review by sdirix — CHANGES_REQUESTED** [review]
Looks good!

**Review by sdirix — APPROVED** [review]

**Pull request merged**